### PR TITLE
Fix bugs in node watcher and add comments to controllers

### PR DESCRIPTION
**Pull request by zionwu**
1. fix issue https://github.com/rancher/rancher/issues/11608
2. add comments to controller.

### CODE REVIEW

File: pkg/controllers/user/alert/watcher/node.go [hunk 1]
```diff
@@ -69,13 +69,8 @@ func (w *NodeWatcher) watchRule() error {
 		if alert.Spec.TargetNode.NodeName != "" {
 			parts := strings.Split(alert.Spec.TargetNode.NodeName, ":")
 			id := parts[1]
-			newNode, err := w.nodeLister.Get("", id)
```
Comment (aiwantaozi) [inline pkg/controllers/user/alert/watcher/node.go]: may be panic, if there are invalid format and the array length is 0.

Comment (aiwantaozi) [inline pkg/controllers/user/alert/watcher/node.go]: id := parts[1]
may be panic, if there are invalid format and the array length is 0.

Comment (zionwu) [reply pkg/controllers/user/alert/watcher/node.go]: I added the following logic to check:
```
if len(parts) != 2 {
  continue
}
```

File: pkg/controllers/user/alert/watcher/node.go [hunk 2]
```diff
@@ -154,13 +172,13 @@ func (w *NodeWatcher) checkNodeCPUUsage(node *corev1.Node, alert *v3.ClusterAler
...
-	for _, cond := range node.Status.Conditions {
+	for _, cond := range machine.Status.InternalNodeStatus.Conditions {
 		if cond.Type == corev1.NodeReady {
 			if cond.Status == corev1.ConditionFalse {
 
-				title := fmt.Sprintf("The kubelet on the node %s is not healthy", node.Name)
+				title := fmt.Sprintf("The kubelet on the node %s is not healthy", getNodeName(machine))
 				desc := fmt.Sprintf("*Alert Name*: %s\n*Cluster Name*: %s\n*Logs*: %s", alert.Spec.DisplayName, w.clusterName, cond.Message)
 
 				if err := w.alertManager.SendAlert(alertID, desc, title, alert.Spec.Severity); err != nil {
```
Comment (aiwantaozi) [inline pkg/controllers/user/alert/watcher/node.go]: is that ok we only use debug log？

**Comment by aiwantaozi** [comment]
LGTM

File: pkg/controllers/user/alert/watcher/deployment.go [hunk 3]
```diff
@@ -87,7 +87,10 @@ func (w *DeploymentWatcher) watchRule() error {
 
 				selector := labels.NewSelector()
 				for key, value := range alert.Spec.TargetWorkload.Selector {
-					r, _ := labels.NewRequirement(key, selection.Equals, []string{value})
+					r, err := labels.NewRequirement(key, selection.Equals, []string{value})
+					if err != nil {
+						continue
```
Comment (alena1108) [inline pkg/controllers/user/alert/watcher/deployment.go]: should we log the error? in warn

File: pkg/controllers/user/alert/watcher/node.go [hunk 4]
```diff
@@ -68,20 +68,21 @@ func (w *NodeWatcher) watchRule() error {
...
+			w.checkNodeCondition(alert, machine)
 
 		} else if alert.Spec.TargetNode.Selector != nil {
 
 			selector := labels.NewSelector()
 			for key, value := range alert.Spec.TargetNode.Selector {
-				r, _ := labels.NewRequirement(key, selection.Equals, []string{value})
+				r, err := labels.NewRequirement(key, selection.Equals, []string{value})
+				if err != nil {
+					continue
```
Comment (alena1108) [inline pkg/controllers/user/alert/watcher/node.go]: same here, do you think we can benefit from logging the error? in Warn

File: pkg/controllers/user/alert/watcher/node.go [hunk 5]
```diff
@@ -90,43 +91,63 @@ func (w *NodeWatcher) watchRule() error {
...
+func getMachineByNodeName(machines []*v3.Node, nodeName string) *v3.Node {
+	for _, m := range machines {
+		if getNodeName(m) == nodeName {
+			return m
+		}
+	}
 	return nil
+}
 
+func getNodeName(machine *v3.Node) string {
```
Comment (alena1108) [inline pkg/controllers/user/alert/watcher/node.go]: this code looks familiar  - we use something similar in provisioner. Lets think about putting it to some kind of a common file, and use from both controllers.

File: pkg/controllers/user/alert/watcher/deployment.go (hunk 3 above)
Comment (zionwu) [reply pkg/controllers/user/alert/watcher/deployment.go]: ok

File: pkg/controllers/user/alert/watcher/node.go (hunk 5 above)
Comment (zionwu) [reply pkg/controllers/user/alert/watcher/node.go]: ok

**Pull request merged**